quint/src/quintIr.ts

```diff
@@ -248,10 +248,43 @@ export interface QuintLet extends WithId {
   expr: QuintEx
 }
 
+export type MatchCase = {
+  /** the variant label to match */
+  label: string
+  /** the lamba to apply to the variant's wrapped value */
+  elim: QuintLambda
+}
+
+export interface QuintMatch extends WithId {
+  kind: 'match'
+  /** the variant expression to match */
+  expr: QuintEx
+  /** the cases to match against */
+  cases: MatchCase[]
+}
+
+/** A variant value inhabiting a sum type */
+export interface QuintVariant extends WithId {
+  kind: 'variant'
+  /** the varian't tag, e.g., the `A` in `A(n)` */
+  label: string
+  /** the expression injected into the sum type  */
+  expr?: QuintEx
+}
+
```

Comment:
In general:

> we have managed to keep the number of cases in QuintEx at minimum. This required us to introduce "special" low-level operators like Rec(...) and field(...) instead of having distinct cases in the IR. Although the normal users should not use these operators, they simplify IR manipulation at large.

But what is the gain here? I've enumerated some of the costs here:

- https://github.com/informalsystems/quint/blob/65f4877aba142b3c7f5538920bb568a69c3f8244/doc/rfcs/rfc001-sum-types/rfc001-sum-types.org#drop-the-exotic-operators
- https://github.com/informalsystems/quint/issues/1081

In the case of these kind of special constructs, any savings on the cases on `QuintEx` are paid for directly in the number of cases in the operator application logic, only we lose structure along the way and up having to perform unnecessarily validations and computations to reconstruct that lost data.

This PR is meant in part to show the benefit of dropping these exotic operator in favor a richer and more helpful representation in the IR.

But if you could help me understand where we actually see simplifications instead of complexity, that could help me move bast my aversion here.

In particular:

Wouldn't adding the `varaiant` operator require support for open rows in sum-types, which we agreed (as per your preference) we did not want to expose in #1062?

We will need special cases for most operator applications to handle `match`, but we will lose the typing info and structure provided here.